File: modules/dxp/apps/osb/osb-asah/osb-asah-stream-curator/src/main/java/com/liferay/osb/asah/stream/curator/model/blog/Blog.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Liferay Enterprise
 * Subscription License ("License"). You may not use this file except in
 * compliance with the License. You can obtain a copy of the License by
 * contacting Liferay, Inc. See the License for the specific language governing
 * permissions and limitations under the License, including but not limited to
 * distribution rights of the Software.
 *
 *
 *
 */

package com.liferay.osb.asah.stream.curator.model.blog;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import com.liferay.osb.asah.common.date.DateUtil;
import com.liferay.osb.asah.stream.curator.model.BaseAssetModel;

import java.util.Date;

/**
 * @author Inácio Nery
 */
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonTypeInfo(
	defaultImpl = Blog.class, include = JsonTypeInfo.As.EXISTING_PROPERTY,
	property = "messageFormat", use = JsonTypeInfo.Id.NAME, visible = true
)
public class Blog extends BaseAssetModel {

	public void addClicks(long clicks) {
		_clicks += clicks;
	}

	public void addComments(long comments) {
		_comments += comments;
	}

	public void addViews(long views) {
		_views += views;
	}

	public long getClicks() {
		return _clicks;
	}

	public long getComments() {
		return _comments;
	}

	@JsonFormat(
		pattern = DateUtil.PATTERN_ISO_8601, shape = JsonFormat.Shape.STRING,
		timezone = "UTC"
	)
	public Date getFirstEventDate() {
		if (_firstEventDate == null) {
			return null;
		}

		return new Date(_firstEventDate.getTime());
	}

	public long getRatings() {
		return _ratings;
	}

	public float getRatingsScore() {
		return _ratingsScore;
	}

	public int getReadPercentile() {
		return _readPercentile;
	}

	public long getReadTime() {
		return _readTime;
	}

	public long getViews() {
		return _views;
	}

	public void setClicks(long clicks) {
		_clicks = clicks;
	}

	public void setComments(long comments) {
		_comments = comments;
	}

	public void setFirstEventDate(Date firstEventDate) {
		if (firstEventDate != null) {
			_firstEventDate = new Date(firstEventDate.getTime());
		}
	}

	public void setRatings(long ratings) {
		_ratings = ratings;
	}

	public void setRatingsScore(float ratingsScore) {
		_ratingsScore = ratingsScore;
	}

	public void setReadPercentile(int readPercentile) {
		_readPercentile = readPercentile;
	}

	public void setReadTime(long readTime) {
		_readTime = readTime;
	}

	public void setViews(long views) {
		_views = views;
	}

	private long _clicks;
	private long _comments;
	private Date _firstEventDate;
	private long _ratings;
	private float _ratingsScore;
	private int _readPercentile;
	private long _readTime;
	private long _views;

}

File: modules/dxp/apps/osb/osb-asah/osb-asah-spark/src/main/python/liferay/stream/processor/blog.py
#
# Copyright (c) 2000-present Liferay, Inc. All rights reserved.
#
# The contents of this file are subject to the terms of the Liferay Enterprise
# Subscription License ("License"). You may not use this file except in
# compliance with the License. You can obtain a copy of the License by
# contacting Liferay, Inc. See the License for the specific language governing
# permissions and limitations under the License, including but not limited to
# distribution rights of the Software.
#

from liferay.stream.processor.common import AnalyticsEventsDataFrameProcessor

from pyspark.sql import Window, \
	functions as F

class BlogDataFrameProcessor(AnalyticsEventsDataFrameProcessor):

	def _calculate_read_time(self, data_frame):
		data_frame = data_frame.filter("applicationId = 'Blog'")

		window = Window.partitionBy(
			'projectId', 'channelId', 'userId', 'assetId', 'variantId'
		).orderBy(
			F.asc('eventDate')
		).rowsBetween(
			Window.unboundedPreceding, Window.currentRow - 1
		)

		data_frame = data_frame.withColumn(
			'previous_blog_viewed_event_date',
			F.max(
				F.when(
					F.col('eventId') == 'blogViewed',
					F.col('event_date')
				)
			).over(window)
		)

		data_frame = data_frame.filter(F.col('eventId') != 'blogViewed')

		data_frame = data_frame.withColumn(
			'previous_blog_viewed_to_current_event_delta',
			F.col('event_date').cast('long') -
			F.col('previous_blog_viewed_event_date').cast('long')
		)

		window = Window.partitionBy(
			'projectId', 'channelId', 'userId', 'assetId', 'variantId',
			'previous_blog_viewed_event_date'
		).orderBy(
			F.desc('previous_blog_viewed_to_current_event_delta')
		)

		data_frame = data_frame.withColumn(
			'row_number',
			F.row_number().over(window)
		)

		data_frame = data_frame.filter('row_number = 1')

		data_frame = data_frame.withColumn(
			'last_blog_event_date',
			F.max(
				F.col('event_date')
			).over(
				Window.partitionBy(
					'projectId', 'channelId', 'userId', 'assetId', 'variantId'
				)
			)
		)

		return data_frame.groupby(
			'projectId', 'channelId', 'userId', 'assetId', 'variantId',
			'last_blog_event_date'
		).sum(
			'previous_blog_viewed_to_current_event_delta'
		).withColumnRenamed(
			'last_blog_event_date', 'event_date'
		).withColumnRenamed(
			'sum(previous_blog_viewed_to_current_event_delta)',
			'read_time'
		)

	def _filter(self, analytics_events_data_frame):
		return analytics_events_data_frame.filter(
			"""
				(
					(
						(applicationId = 'Comment') AND 
						(eventId = 'posted') AND 
						(eventProperties.className = '{class_name}') AND
						(eventProperties.classPK != '')
					) OR
					(
						(applicationId = 'Blog') AND 
						(
							(eventId = 'blogClicked') OR
							(eventId = 'blogDepthReached') OR
							(eventId = 'blogViewed')
						) AND 
						( 
							(eventProperties.entryId != '')
						)
					) OR 
					(
						(applicationId = 'Ratings') AND 
						(eventId = 'VOTE') AND
						(eventProperties.className = '{class_name}') AND
						(eventProperties.classPK != '') AND
						(
							(eventProperties.ratingType IS NULL) OR
							(eventProperties.ratingType = 'stars')
						)
					)
				)	
			""".format(
				class_name='com.liferay.blogs.model.BlogsEntry'
			)
		)

	def _get_asset_id_column(self):
		return F.when(
			F.col('eventId').isin('posted', 'VOTE'),
			F.col('eventProperties.classPK')
		).otherwise(
			F.col('eventProperties.entryId')
		)

	def _process(self, data_frame):
		data_frame_with_rating_score = self._calculate_rating_score(
			data_frame
		)

		data_frame = data_frame.filter(
			"applicationId != 'Ratings'"
		).withColumn(
			'clicks',
			F.when(
				F.col('eventId') == 'blogClicked', F.lit(1)
			).otherwise(
				F.lit(0)
			)
		).withColumn(
			'comments',
			F.when(
				F.col('eventId') == 'posted', F.lit(1)
			).otherwise(
				F.lit(0)
			)
		).withColumn(
			'views',
			F.when(
				F.col('eventId') == 'blogViewed', F.lit(1)
			).otherwise(
				F.lit(0)
			)
		).groupBy(
			'projectId', 'channelId', 'userId', 'assetId', 'variantId',
			'normalized_event_date', 'primaryKey'
		).agg(
			F.sum('clicks').alias('clicks'),
			F.sum('comments').alias('comments'),
			F.sum('views').alias('views')
		)

		return data_frame.join(
			data_frame_with_rating_score,
			how='left',
			on=[
				'projectId', 'channelId', 'userId', 'assetId', 'variantId',
				'normalized_event_date', 'primaryKey'
			]
		).fillna(0)